stimproc/testing/netstimtester.py

```diff
@@ -0,0 +1,260 @@
+
+            if n > max:
+                max = n
+
+        logging.info(
+            "Min: " + str(round(min, 3)) + " ms\t\t\t\tMax: " + str(round(max, 3)) + " ms\t\t\t\tAvg: " + str(round(avg, 3)) + " ms")
+
+    # Writes csv file
+    def writeCSV(self):
+        with open(f'csv/{self.initTime}.csv', 'w', newline='') as csvfile:
+            f = csv.writer(csvfile)
+            f.writerows(self.testCSV)
+
+    # Creates a line split on the console
+    def lineSplit(self):
+        logging.info("\n")
+        logging.info("-" * 75)
+        logging.info("\n")
+
+    # Prints colored text to console
+    def printResult(self, text, isPass, isBold):
+        if isPass:
+            print('\033[92m', end="")
+        else:
+            print('\033[91m', end="")
+        if isBold:
+            print('\033[1m', end="")
+
+        logging.info(text)
+        print('\033[0m', end="")
+
+    # Runs all tests
+    def runTests(self):
+
+        for test in self.tests:  # iterates through all tests
+
+            # establishes server connection
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+                s.bind((self.HOST, self.PORT))
+                s.listen()
+                self.conn, self.addr = s.accept()
+
+                with self.conn:
+                    self.runCommands(test)
+
+            sleep(0.001)  # millisecond delay prevents connection from failing - keep this in
+
+    # Runs all input commands
+    def runCommands(self, commands):
+
+        testPass = True
+
+        for command in commands:  # executes each command
+
+            logging.info("Send: " + command[0].decode())
+            self.conn.sendall(command[0] + b'\n')  # sends command to stimproc
+
+            # Timer measures time between sent command and received data
+            start = time()
+            data = self.conn.recv(1024)
+            end = time()
+            responseTime = (end - start) * 1000  # calculates response time in ms rounded to 3 decimals
+
+            logging.info("Recv: " + data.decode().strip())
+            logging.info("Response Time: " + str(round(responseTime, 3)) + " ms")
+            logging.info("Threshold: " + str(command[1]) + " ms")
+
+            # Determines if response time measured is within threshold
+            if responseTime <= command[1]:
+                withinThreshold = True
+
+            else:
+                withinThreshold = False
+
+            logging.info("Within Threshold: " + str(withinThreshold))
+
+            # Determines if response matches the expected response
+            if data.strip().split(b',')[0] == command[2]:
+                expectedResponse = True
+            else:
+                expectedResponse = False
+
+            logging.info("Expected Response: " + str(expectedResponse))
+
+            # Determines if the test is a pass or failure - note that expected errors constitute passes
+            if withinThreshold and expectedResponse:
```

Comment:
Statistical values for response times must include responses which are not withinThreshold so that it is properly reporting issues with unexpectedly long delays.